Adds GoogleSQL Syntax Highlighting

pratik-m opened the pull request:
<!--
The manual review may take several days or weeks,
depending on the reviewer's availability and workload.
Patience padawan!

You can request a review from @packagecontrol-bot.
Please ensure the reviews pass and follow any instructions.

Please provide some information via this checklist,
feel free to remove what't not applicable.
-->

- [x ] I'm the package's author and/or maintainer.
- [ x] I have have read [the docs][1].
- [ x] I have tagged a release with a [semver][2] version number.
- [ x] My package repo has a description and a README describing what it's for and how to use it.
- [x] My package doesn't add context menu entries. *
- [x] My package doesn't add key bindings. **
- [x] Any commands are available via the command palette.
- [ ] Preferences and keybindings (if any) are listed in the menu and the command palette, and open in split view.
- [ x] If my package is a syntax it doesn't also add a color scheme. ***
- [ x] If my package is a syntax it is named after the language it supports (without suffixes like "syntax" or "highlighting").
- [ ] I use [.gitattributes][3] to exclude files from the package: images, test files, sublime-project/workspace.

My package is GoogleSQL Syntax Highlighting

There are no packages like it in Package Control.

<!-- 
*)   Unless it definitely really needs them,
     they apply to the cursor's context
     and their visibility is conditional.
     Space in this menu is limited!
**)  There aren't enough keys for all packages,
     you'd risk overriding those of other packages.
     You can put commented out suggestions in a keymap file, 
     and/or explain how to create bindings in your README.
***) We have hundreds of color schemes,
     and plenty of scopes to make any syntax work. 

For bonus points also considered how the review guidelines apply to your package:
https://github.com/wbond/package_control_channel/wiki#reviewing-a-package-addition

For updates to existing packages:
If your package isn't using tag based releases,
please switch to tags now.
 -->

[1]: https://packagecontrol.io/docs/submitting_a_package
[2]: https://semver.org
[3]: https://www.git-scm.com/docs/gitattributes#_export_ignore

braver commented on repository/g.json:
```diff
@@ -1590,6 +1590,18 @@
 			]
 		},
+		{
+			"name": "GoogleSQL Syntax Highlighting",
```
```suggestion
			"name": "GoogleSQL",
```

> If my package is a syntax it is named after the language it supports (without suffixes like "syntax" or "highlighting").

braver commented on repository/g.json:
```diff
@@ -1590,6 +1590,18 @@
+			"name": "GoogleSQL Syntax Highlighting",
+			"details": "https://github.com/pratik-m/sublime-googlesql-syntax",
+			"author": "Pratik Munot",
+			"labels": ["syntax", "GoogleSQL", "BigQuery"],
```
```suggestion
			"labels": ["language syntax", "GoogleSQL", "BigQuery"],
```

https://packagecontrol.io/browse/labels/language%20syntax

braver commented on repository/g.json:
```diff
@@ -1590,6 +1590,18 @@
+			"releases": [
+				{
+					"sublime_text": "*",
+					"branch": "main"
```
Branch based releases are ancient history.

Also
> I have tagged a release with a [semver](https://semver.org/) version number.

.. you didn't actually.

pratik-m replied:
Added tags and created a release with semver version number.

pratik-m commented:
Thanks @braver ! I've made the changes based on your feedback!

braver commented:
Excellent, thanks 👍🏻

braver commented:
I have some remarks about the syntax itself, will get back to you with that!

braver commented:
- Since you register your syntax to handle `.sql` files, you'll probably interfere with the existing SQL language package. It's better to leave that more generic extension alone, focusing on a specific extension for your syntax (hopefully there is one) and then tell users how to switch if they do want to use GoogleSQL syntax for all sql files.
- Similarly you use the `source.sql` main scope and `.sql` suffix on each scope. Since your syntax if probably very much _like_ SQL but not _exactly_ SQL, it's better to have a suffix of your own. An example would be Twig, which is mostly just html, but with its own specifics on top, so it uses the `text.html.twig` scope (ie `.twig` on top of `text.html`). That way selectors targeting HTML would match, but the specific Twig language things can also be targeted individually. You might not want that if GoogleSQL is too different from SQL, in which case `source.googlesql` might be better (ie. a completely different suffix).
- I wonder if you could extend the default SQL syntax and add GoogleSQL-specific adjustments on top? That might be easier to maintain over time. I do say that without knowing anything about GoogleSQL though 😅

pratik-m commented:
Hi @braver 
As you mentioned, GoogleSQL is very similar to standard SQL, with a few additional keywords and syntax variations. While Google refers to the SQL dialect used on their platform as "GoogleSQL," I don't believe there's a dedicated extension specifically for it.
- I’ve created a syntax file to support the GoogleSQL-specific syntax, using standard SQL as the base.
- I’m not sure introducing a separate file extension would be helpful, since most users will likely continue saving their files with the .sql extension.
- Could you guide me on how to update the default SQL syntax to incorporate the GoogleSQL additions? I tried looking through Sublime’s GitHub repository but couldn’t find a clear approach.

That said, I still think it might be worth keeping the GoogleSQL syntax highlighter separate, so we have flexibility to accommodate any future changes.

Would love to hear your thoughts or suggestions.

braver commented:
A reason to not specify the .sql extension is to make sure you don't "steal" it.. but then again, If someone is installing your package, that's likely to be what they want it to do. It might not work anyway, I'm not 100% how Sublime decides which syntax to apply in that case where there are multiple for same extension... it might pick the first match, or it might pick the last match. Anyway, that's fine by me either way.

So, the idea would be to introduce your new syntax alongside the existing SQL syntax. But you can use that syntax as a basis (ie. extend it) so you only have to add what's different in GoogleSQL: https://www.sublimetext.com/docs/syntax.html#inheritance

pratik-m commented:
Hi @braver - I think I can extend the default SQL syntax and override the elements for BigQuery. I've started working on it and will share the updates in few days. I'll retain the extension to '.sql'..
Thanks!

pratik-m commented:
Hi @braver  - As per your suggestion, I've extended the delivered templates and have overriden the tags specific to BigQuery. 
- I've simplified the yaml as well by consolidating all the functions into one, instead of having categories for them. 
- Added a three GoogleSQL related syntax  
 Can you pls review it and let me know your thoughts? 
 
 Thanks Pratik

braver commented:
Nice job!

pratik-m commented:
Hi @braver - My package is not coming up on https://packagecontrol.io/ 
I tried to some digging around but couldn't pinpoint the issue. Can you help me point to the issue and I'll get it fixed? 
Thanks!

braver commented:
It’s probably just the index not updating properly, we’ve been having intermittent issues with that for a while. Nothing you or I can do something about, people who can are already working on it. Just give it some time 🤞🏻

pratik-m commented:
Thanks for the update @braver. I'll keep an eye out!
